File: src/main/java/me/partlysanestudios/partlysaneskies/system/ThemeManager.java
//
// Written by Su386.
// See LICENSE for copyright and license notices.
//

package me.partlysanestudios.partlysaneskies.system;

import cc.polyfrost.oneconfig.config.core.OneColor;
import gg.essential.elementa.components.UIImage;
import gg.essential.universal.utils.ReleasedDynamicTexture;
import me.partlysanestudios.partlysaneskies.PartlySaneSkies;
import me.partlysanestudios.partlysaneskies.utils.ElementaUtils;
import me.partlysanestudios.partlysaneskies.utils.ImageUtils;
import net.minecraft.client.renderer.texture.TextureUtil;
import net.minecraft.client.resources.IResource;
import net.minecraft.util.ResourceLocation;

import java.awt.*;
import java.awt.image.BufferedImage;
import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;

public class ThemeManager {

    private static final Color PRIMARY_DEBUG_COLOR = new Color(255, 0, 0);
    private static final Color SECONDARY_DEBUG_COLOR = new Color(0, 0, 255);
    private static final Color ACCENT_DEBUG_COLOR = new Color(0, 255, 0);
    private static String lastThemeName = "";

    private static final ArrayList<UIImage> backgroundUIImages = new ArrayList<>();
    private static final ArrayList<ButtonData> buttonDataList = new ArrayList<>();
    private static final ArrayList<ToggleData> toggleDataList = new ArrayList<>();

    public static Theme[] defaultThemes = {
            new Theme("Classic (Dark)",  new Color(46, 47, 50), new Color(37, 38, 41), new Color(1, 255, 255)),
            new Theme("Royal Dark (Dark)", new Color(46, 47, 50), new Color(39, 31, 43), new Color(91, 192, 235)),
            new Theme("Midnight Forest (Dark)", new Color(46, 47, 50),  new Color(40, 50, 38), new Color(35, 206, 107)),

            new Theme("Moonless Night (Very Dark)", new Color(24, 24, 27), new Color(15, 17, 20), new Color(8, 124, 167)),
            new Theme("Stormy Night (Very Dark)", new Color(23, 23, 34), new Color(5, 5, 27), new Color(94, 74, 227)),
            new Theme("The Void (Very Dark)", new Color(14, 17, 21), new Color(5, 5, 12), new Color(113, 179, 64)),

            new Theme("Classic (Light)", new Color(245, 245, 245), new Color(213, 210, 195), new Color(42, 84, 209)),
            new Theme("Royal Light (Light)",  new Color(245, 245, 245), new Color(127, 106, 147), new Color(242, 97, 87)),
            new Theme("Partly Cloudy (Light)", new Color(245, 245, 245), new Color(84, 95, 117), new Color(217, 114, 255)),

            new Theme("Waterfall (Colorful)", new Color(214, 237, 246), new Color(172, 215, 236), new Color(108, 197, 81)),
            new Theme("Jungle (Colorful)", new Color(201, 227, 172), new Color(144, 190, 109), new Color(254, 100, 163)),
            new Theme("Dunes (Colorful)", new Color(229, 177, 129), new Color(222, 107, 72), new Color(131, 34, 50))
    };

    public static void run() {
        if (PartlySaneSkies.config.useDefaultAccentColor) {
            PartlySaneSkies.config.accentColor = getAccentColor();
        }

        if (!PartlySaneSkies.config.customTheme) {
            PartlySaneSkies.config.primaryColor = getPrimaryColor();
            PartlySaneSkies.config.secondaryColor = getSecondaryColor();
        }

//        If the theme has changed
        if (!lastThemeName.equals(defaultThemes[PartlySaneSkies.config.themeIndex].getName())) {
            for (UIImage image : backgroundUIImages) {
                try {
                    image.applyTexture(new ReleasedDynamicTexture(ImageUtils.INSTANCE.loadImage(getCurrentBackgroundFile().getPath())));
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }

            for (ButtonData data : buttonDataList) {
                try {
                    OneColor color;
                    if (data.getColor() == null) {
                        color = getAccentColor();
                    } else {
                        color = data.getColor();
                    }

                    data.getImage().applyTexture(new ReleasedDynamicTexture(ImageUtils.INSTANCE.loadImage(getCurrentButtonFile(color).getPath())));
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }

            for (ToggleData data : toggleDataList) {
                try {
                    OneColor color;
                    if (data.getColor() == null) {
                        color = getAccentColor();
                    } else {
                        color = data.getColor();
                    }

                    data.getImage().applyTexture(new ReleasedDynamicTexture(ImageUtils.INSTANCE.loadImage(getCurrentToggleFile(data.isSelected(), color).getPath())));
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }

            lastThemeName = defaultThemes[PartlySaneSkies.config.themeIndex].getName();
        }


    }


    public static UIImage getCurrentBackgroundUIImage() {
        UIImage image;
        if (PartlySaneSkies.config.disableThemes) {
            image = ElementaUtils.INSTANCE.uiImageFromResourceLocation(new ResourceLocation("partlysaneskies", "textures/gui/base_color_background.png"));

        }
        else {
            try {
                image = UIImage.ofFile(getCurrentBackgroundFile());
            } catch (IOException e) {
                image = ElementaUtils.INSTANCE.uiImageFromResourceLocation(new ResourceLocation("partlysaneskies", "textures/gui/base_color_background.png"));
            }
        }

        backgroundUIImages.add(image);
        return image;
    }

    public static UIImage getCurrentButtonUIImage() {
        return getCurrentButtonUIImage(getAccentColor());
    }

    public static UIImage getCurrentButtonUIImage(OneColor accentColor) {
        UIImage image;
        if (PartlySaneSkies.config.disableThemes) {
            if (accentColor.equals(getAccentColor())) {
                image = ElementaUtils.INSTANCE.uiImageFromResourceLocation(new ResourceLocation("partlysaneskies", "textures/gui/base_color_button.png"));
            } else {
                image = ElementaUtils.INSTANCE.uiImageFromResourceLocation(new ResourceLocation("partlysaneskies", "textures/gui/base_color_button_transparent.png"));
            }
        } else {
            try {
                image = UIImage.ofFile(getCurrentButtonFile(accentColor));
            } catch (IOException e) {
                if (accentColor.equals(getAccentColor())) {
                    image = ElementaUtils.INSTANCE.uiImageFromResourceLocation(new ResourceLocation("partlysaneskies", "textures/gui/base_color_button.png"));
                } else {
                    image = ElementaUtils.INSTANCE.uiImageFromResourceLocation(new ResourceLocation("partlysaneskies", "textures/gui/base_color_button_transparent.png"));
                }
            }
        }

        if (accentColor.equals(getAccentColor())) {
            accentColor = null;
        }

        buttonDataList.add(new ButtonData(image, accentColor));
        return image;
    }


    public static UIImage getCurrentToggleUIImage(boolean selected) {
        return getCurrentToggleUIImage(selected, getAccentColor());
    }

    public static UIImage getCurrentToggleUIImage(boolean selected, OneColor accentColor) {
        UIImage image;
        if (PartlySaneSkies.config.disableThemes) {
            if (selected) {
                image = ElementaUtils.INSTANCE.uiImageFromResourceLocation(new ResourceLocation("partlysaneskies" , "textures/gui/selected_toggle.png"));
            } else {
                image = ElementaUtils.INSTANCE.uiImageFromResourceLocation(new ResourceLocation("partlysaneskies" , "textures/gui/unselected_toggle.png"));
            }
        }
        try {
            image = UIImage.ofFile(getCurrentToggleFile(selected, accentColor));
        } catch (IOException e) {
            if (selected) {
                image = ElementaUtils.INSTANCE.uiImageFromResourceLocation(new ResourceLocation("partlysaneskies" , "textures/gui/selected_toggle.png"));
            } else {
                image = ElementaUtils.INSTANCE.uiImageFromResourceLocation(new ResourceLocation("partlysaneskies" , "textures/gui/unselected_toggle.png"));
            }
        }

        if (accentColor.equals(getAccentColor())) {
            accentColor = null;
        }

        toggleDataList.add(new ToggleData(image, selected, accentColor));
        return image;
    }

    public static File getCurrentBackgroundFile() throws IOException {
        return getBackgroundWithColor(getPrimaryColor(), getSecondaryColor(), getAccentColor());
    }

    public static File getCurrentButtonFile() throws IOException {
        return getCurrentButtonFile(getAccentColor());
    }

    public static File getCurrentButtonFile(OneColor accentColor) throws IOException {
        return getButtonWithColor(getPrimaryColor(), getSecondaryColor(), accentColor);
    }

    public static File getCurrentToggleFile(boolean selected) throws IOException {
        return getCurrentToggleFile(selected, getAccentColor());
    }

    public static File getCurrentToggleFile(boolean selected, OneColor accentColor) throws IOException {
        if (selected) {
            return getToggleWithColor(getPrimaryColor(), getSecondaryColor(), accentColor);
        } else {
            return getToggleWithColor(getPrimaryColor(), getSecondaryColor(), getSecondaryColor());
        }
    }

    public static OneColor getPrimaryColor() {
        if (!PartlySaneSkies.config.customTheme) {
            int themeIndex = PartlySaneSkies.config.themeIndex;
            return new OneColor(defaultThemes[themeIndex].getPrimaryColor());
        } else {
            return PartlySaneSkies.config.primaryColor;
        }
    }

    public static OneColor getDarkPrimaryColor() {
        return new OneColor(darkenColor(getPrimaryColor()));
    }

    public static OneColor getLightPrimaryColor() {
        return new OneColor(lightenColor(getPrimaryColor()));
    }

    public static OneColor getSecondaryColor() {
        if (!PartlySaneSkies.config.customTheme) {
            int themeIndex = PartlySaneSkies.config.themeIndex;
            return new OneColor(defaultThemes[themeIndex].getSecondaryColor());
        } else {
            return PartlySaneSkies.config.secondaryColor;
        }
    }

    public static OneColor getDarkSecondaryColor() {
        return new OneColor(darkenColor(getSecondaryColor()));
    }

    public static OneColor getLightSecondaryColor() {
        return new OneColor(lightenColor(getSecondaryColor()));
    }

    public static OneColor getAccentColor() {
        if (PartlySaneSkies.config.useDefaultAccentColor) {
            int themeIndex = PartlySaneSkies.config.themeIndex;
            return new OneColor(defaultThemes[themeIndex].getDefaultAccentColor());
        } else {
            return PartlySaneSkies.config.accentColor;
        }
    }

    public static Color getDarkAccentColor() {
        return darkenColor(getAccentColor());
    }

    public static Color getLightAccentColor() {
        return lightenColor(getAccentColor());
    }

    private static Color darkenColor(OneColor color) {
        return darkenColor(color.toJavaColor());
    }
    private static Color darkenColor(Color color) {
        int averageR = (int) (color.getRed()* .761);
        int averageG = (int) (color.getGreen() * .761);
        int averageB = (int) (color.getBlue() * .761);

        return new Color(averageR, averageG, averageB, color.getTransparency());
    }


    private static Color lightenColor(OneColor color) {
        return lightenColor(color.toJavaColor());
    }

    private static Color lightenColor(Color color) {
        int averageR = (int) (color.getRed()* .798 + 255 * .202);
        int averageG = (int) (color.getGreen() * .798 + 255 * .202);
        int averageB = (int) (color.getBlue() * .798 + 255 * .202);

        return new Color(averageR, averageG, averageB, color.getTransparency());
    }

    public static File getBackgroundWithColor(OneColor primaryColor, OneColor secondaryColor, OneColor accentColor) throws IOException {


        String primaryColorHex = Integer.toHexString(primaryColor.getRGB() & 0xffffff);
        String secondaryColorHex = Integer.toHexString(secondaryColor.getRGB() & 0xffffff);
        String accentColorHex = Integer.toHexString(accentColor.getRGB() & 0xffffff);

        Path folderPath = Paths.get("./config/partly-sane-skies/image_variants/background");
        String fileName = "background-" + primaryColorHex + "-" + secondaryColorHex + "-" + accentColorHex + ".png";
        Path filePath = Paths.get(folderPath + "/" + fileName);

        if (Files.exists(filePath)){
            return filePath.toFile();
        }


        ResourceLocation backgroundResourceLocation = new ResourceLocation("partlysaneskies", "textures/debug_gui_textures/background.png");

        IResource debugTexture = PartlySaneSkies.minecraft.getResourceManager().getResource(backgroundResourceLocation);
        BufferedImage debugImage = TextureUtil.readBufferedImage(debugTexture.getInputStream());

        folderPath.toFile().mkdirs();
        filePath.toFile().createNewFile();


        ImageUtils.INSTANCE.replaceColor(debugImage, PRIMARY_DEBUG_COLOR, primaryColor);
        ImageUtils.INSTANCE.replaceColor(debugImage, SECONDARY_DEBUG_COLOR, secondaryColor);
        ImageUtils.INSTANCE.replaceColor(debugImage, ACCENT_DEBUG_COLOR, accentColor);

        ImageUtils.INSTANCE.saveImage(debugImage, filePath);



        return filePath.toFile();
    }

    public static File getButtonWithColor(OneColor primaryColor, OneColor secondaryColor, OneColor accentColor) throws IOException {


        String primaryHex = Integer.toHexString(primaryColor.getRGB() & 0xffffff);
        String secondaryColorHex = Integer.toHexString(secondaryColor.getRGB() & 0xffffff);
        String accentColorHex = Integer.toHexString(accentColor.getRGB() & 0xffffff);

        Path folderPath = Paths.get("./config/partly-sane-skies/image_variants/button");
        String fileName = "button-" + primaryHex + "-" + secondaryColorHex + "-" + accentColorHex + ".png";
        Path filePath = Paths.get(folderPath + "/" + fileName);

        if (Files.exists(filePath)){
            return filePath.toFile();
        }


        ResourceLocation buttonResourceLocation = new ResourceLocation("partlysaneskies", "textures/debug_gui_textures/button.png");

        IResource debugTexture = PartlySaneSkies.minecraft.getResourceManager().getResource(buttonResourceLocation);
        BufferedImage debugImage = TextureUtil.readBufferedImage(debugTexture.getInputStream());

        folderPath.toFile().mkdirs();
        filePath.toFile().createNewFile();


        ImageUtils.INSTANCE.replaceColor(debugImage, PRIMARY_DEBUG_COLOR, primaryColor);
        ImageUtils.INSTANCE.replaceColor(debugImage, SECONDARY_DEBUG_COLOR, secondaryColor);
        ImageUtils.INSTANCE.replaceColor(debugImage, ACCENT_DEBUG_COLOR, accentColor);

        ImageUtils.INSTANCE.saveImage(debugImage, filePath);



        return filePath.toFile();
    }

    public static File getToggleWithColor(OneColor primaryColor, OneColor secondaryColor, OneColor accentColor) throws IOException {


        String primaryColorHex = Integer.toHexString(primaryColor.getRGB() & 0xffffff);
        String secondaryColorHex = Integer.toHexString(secondaryColor.getRGB() & 0xffffff);
        String accentColorHex = Integer.toHexString(accentColor.getRGB() & 0xffffff);

        Path folderPath = Paths.get("./config/partly-sane-skies/image_variants/toggle");
        String fileName = "toggle-" + primaryColorHex + "-" + secondaryColorHex + "-" + accentColorHex + ".png";
        Path filePath = Paths.get(folderPath + "/" + fileName);

        if (Files.exists(filePath)){
            return filePath.toFile();
        }


        ResourceLocation toggleResourceLocation = new ResourceLocation("partlysaneskies", "textures/debug_gui_textures/toggle.png");

        IResource debugTexture = PartlySaneSkies.minecraft.getResourceManager().getResource(toggleResourceLocation);
        BufferedImage debugImage = TextureUtil.readBufferedImage(debugTexture.getInputStream());

        folderPath.toFile().mkdirs();
        filePath.toFile().createNewFile();


        ImageUtils.INSTANCE.replaceColor(debugImage, PRIMARY_DEBUG_COLOR, primaryColor);
        ImageUtils.INSTANCE.replaceColor(debugImage, SECONDARY_DEBUG_COLOR, secondaryColor);
        ImageUtils.INSTANCE.replaceColor(debugImage, ACCENT_DEBUG_COLOR, accentColor);

        ImageUtils.INSTANCE.saveImage(debugImage, filePath);


        return filePath.toFile();
    }

    public static class ButtonData {
        private final UIImage image;
        private final OneColor color;
        public ButtonData(UIImage image) {
            this.image = image;
            this.color = null;
        }
        public ButtonData(UIImage image, OneColor color) {
            this.image = image;
            this.color = color;
        }

        public UIImage getImage() {
            return image;
        }

        public OneColor getColor() {
            return color;
        }
    }

    public static class ToggleData {
        private final UIImage image;
        private final OneColor color;
        private final boolean selected;
        public ToggleData(UIImage image, boolean selected) {
            this.image = image;
            this.color = null;
            this.selected = selected;
        }

        public ToggleData(UIImage image, boolean selected, OneColor color) {
            this.image = image;
            this.color = color;
            this.selected = selected;
        }

        public UIImage getImage() {
            return image;
        }

        public OneColor getColor() {
            return color;
        }

        public boolean isSelected() {
            return selected;
        }
    }


    public static class Theme {
        private final String name;
        private final Color primaryColor;
        private final Color secondaryColor;
        private final Color defaultAccentColor;

        public Theme(String name, Color primaryColor, Color secondaryColor, Color defaultAccentColor) {
            this.name = name;
            this.primaryColor = primaryColor;
            this.secondaryColor = secondaryColor;
            this.defaultAccentColor = defaultAccentColor;
        }

        public String getName() {
            return name;
        }
        public Color getPrimaryColor() {
            return primaryColor;
        }

        public Color getSecondaryColor() {
            return secondaryColor;
        }

        public Color getDefaultAccentColor() {
            return defaultAccentColor;
        }
    }
}

File: src/main/kotlin/me/partlysanestudios/partlysaneskies/features/themes/ThemeManager.kt
//
// Written by Su386.
// See LICENSE for copyright and license notices.
//
package me.partlysanestudios.partlysaneskies.features.themes

import cc.polyfrost.oneconfig.config.core.OneColor
import gg.essential.elementa.components.UIImage
import gg.essential.universal.utils.ReleasedDynamicTexture
import me.partlysanestudios.partlysaneskies.PartlySaneSkies.Companion.config
import me.partlysanestudios.partlysaneskies.PartlySaneSkies.Companion.minecraft
import me.partlysanestudios.partlysaneskies.utils.ElementaUtils.uiImageFromResourceLocation
import me.partlysanestudios.partlysaneskies.utils.ImageUtils.loadImage
import me.partlysanestudios.partlysaneskies.utils.ImageUtils.replaceColor
import me.partlysanestudios.partlysaneskies.utils.ImageUtils.saveImage
import net.minecraft.client.renderer.texture.TextureUtil
import net.minecraft.client.resources.IResource
import net.minecraft.util.ResourceLocation
import java.awt.Color
import java.awt.image.BufferedImage
import java.io.File
import java.io.IOException
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.Paths

object ThemeManager {
    private val PRIMARY_DEBUG_COLOR: Color = Color(255, 0, 0)
    private val SECONDARY_DEBUG_COLOR: Color = Color(0, 0, 255)
    private val ACCENT_DEBUG_COLOR: Color = Color(0, 255, 0)
    private var lastThemeName: String = ""
    private val backgroundUIImages: ArrayList<UIImage> = ArrayList()
    private val buttonDataList: ArrayList<ButtonData> = ArrayList()
    private val toggleDataList: ArrayList<ToggleData> = ArrayList()
    private val defaultThemes: Array<Theme> = arrayOf(
        Theme("Classic (Dark)", Color(46, 47, 50), Color(37, 38, 41), Color(1, 255, 255)),
        Theme("Royal Dark (Dark)", Color(46, 47, 50), Color(39, 31, 43), Color(91, 192, 235)),
        Theme("Midnight Forest (Dark)", Color(46, 47, 50), Color(40, 50, 38), Color(35, 206, 107)),
        Theme("Moonless Night (Very Dark)", Color(24, 24, 27), Color(15, 17, 20), Color(8, 124, 167)),
        Theme("Stormy Night (Very Dark)", Color(23, 23, 34), Color(5, 5, 27), Color(94, 74, 227)),
        Theme("The Void (Very Dark)", Color(14, 17, 21), Color(5, 5, 12), Color(113, 179, 64)),
        Theme("Classic (Light)", Color(245, 245, 245), Color(213, 210, 195), Color(42, 84, 209)),
        Theme("Royal Light (Light)", Color(245, 245, 245), Color(127, 106, 147), Color(242, 97, 87)),
        Theme("Partly Cloudy (Light)", Color(245, 245, 245), Color(84, 95, 117), Color(217, 114, 255)),
        Theme("Waterfall (Colorful)", Color(214, 237, 246), Color(172, 215, 236), Color(108, 197, 81)),
        Theme("Jungle (Colorful)", Color(201, 227, 172), Color(144, 190, 109), Color(254, 100, 163)),
        Theme("Dunes (Colorful)", Color(229, 177, 129), Color(222, 107, 72), Color(131, 34, 50))
    )

    fun tick() {
        if (!config.useCustomAccentColor) {
            config.accentColor = accentColor
        }
        if (!config.customTheme) {
            config.primaryColor = primaryColor
            config.secondaryColor = secondaryColor
        }

//        If the theme has changed
        if (lastThemeName != defaultThemes[config.themeIndex].name) {
            for (image: UIImage in backgroundUIImages) {
                try {
                    image.applyTexture(ReleasedDynamicTexture(loadImage(currentBackgroundFile.path)))
                } catch (e: IOException) {
                    e.printStackTrace()
                }
            }
            for (data: ButtonData in buttonDataList) {
                try {
                    val color = data.color
                    data.image.applyTexture(ReleasedDynamicTexture(loadImage(getCurrentButtonFile(color).path)))
                } catch (e: IOException) {
                    e.printStackTrace()
                }
            }
            for (data: ToggleData in toggleDataList) {
                try {
                    val color = data.color
                    data.image.applyTexture(ReleasedDynamicTexture(loadImage(getCurrentToggleFile(data.isSelected, color).path)))
                } catch (e: IOException) {
                    e.printStackTrace()
                }
            }
            lastThemeName = defaultThemes[config.themeIndex].name
        }
    }

    val currentBackgroundUIImage: UIImage
        get() {
            val image: UIImage = if (config.disableThemes) {
                ResourceLocation(
                    "partlysaneskies",
                    "textures/gui/base_color_background.png"
                ).uiImageFromResourceLocation()
            } else {
                try {
                    UIImage.ofFile(currentBackgroundFile)
                } catch (e: IOException) {
                    ResourceLocation("partlysaneskies","textures/gui/base_color_background.png").uiImageFromResourceLocation()
                }
            }
            backgroundUIImages.add(image)
            return image
        }
    val currentButtonUIImage: UIImage
        get() = getCurrentButtonUIImage(accentColor)

    fun getCurrentButtonUIImage(accentColor: OneColor): UIImage {
        val accentColor: OneColor = accentColor
        val image: UIImage
        if (config.disableThemes) {
            image = if ((accentColor == ThemeManager.accentColor)) {
                ResourceLocation("partlysaneskies","textures/gui/base_color_button.png").uiImageFromResourceLocation()
            } else {
                ResourceLocation("partlysaneskies","textures/gui/base_color_button_transparent.png").uiImageFromResourceLocation()
            }
        } else {
            image = try {
                UIImage.ofFile(getCurrentButtonFile(accentColor))
            } catch (e: IOException) {
                if ((accentColor == ThemeManager.accentColor)) {
                    ResourceLocation("partlysaneskies","textures/gui/base_color_button.png").uiImageFromResourceLocation()
                } else {
                    ResourceLocation("partlysaneskies","textures/gui/base_color_button_transparent.png").uiImageFromResourceLocation()
                }
            }
        }

        buttonDataList.add(ButtonData(image, accentColor))
        return image
    }

    fun getCurrentToggleUIImage(selected: Boolean): UIImage {
        return getCurrentToggleUIImage(selected, accentColor)
    }

    fun getCurrentToggleUIImage(selected: Boolean, accentColor: OneColor): UIImage {
        var accentColor: OneColor = accentColor
        var image: UIImage
        if (config.disableThemes) {
            image = if (selected) {
                ResourceLocation("partlysaneskies","textures/gui/selected_toggle.png").uiImageFromResourceLocation()
            } else {
                ResourceLocation("partlysaneskies","textures/gui/unselected_toggle.png").uiImageFromResourceLocation()
            }
        }
        image = try {
            UIImage.ofFile(getCurrentToggleFile(selected, accentColor))
        } catch (e: IOException) {
            if (selected) {
                ResourceLocation("partlysaneskies","textures/gui/selected_toggle.png").uiImageFromResourceLocation()
            } else {
                ResourceLocation("partlysaneskies","textures/gui/unselected_toggle.png").uiImageFromResourceLocation()
            }
        }

        toggleDataList.add(ToggleData(image, selected, accentColor))
        return image
    }

    val currentBackgroundFile: File
        get() = getBackgroundWithColor(primaryColor, secondaryColor, accentColor)

    val currentButtonFile: File
        get() {
            return getCurrentButtonFile(accentColor)
        }

    fun getCurrentButtonFile(accentColor: OneColor): File {
        return getButtonWithColor(primaryColor, secondaryColor, accentColor)
    }

    fun getCurrentToggleFile(selected: Boolean): File {
        return getCurrentToggleFile(selected, accentColor)
    }

    fun getCurrentToggleFile(selected: Boolean, accentColor: OneColor): File {
        return if (selected) {
            getToggleWithColor(primaryColor, secondaryColor, accentColor)
        } else {
            getToggleWithColor(primaryColor, secondaryColor, secondaryColor)
        }
    }

    val primaryColor: OneColor
        get() {
            return if (!config.customTheme) {
                val themeIndex: Int = config.themeIndex
                OneColor(defaultThemes[themeIndex].primaryColor)
            } else {
                config.primaryColor
            }
        }
    val darkPrimaryColor: OneColor
        get() {
            return OneColor(darkenColor(primaryColor))
        }
    val lightPrimaryColor: OneColor
        get() {
            return OneColor(lightenColor(primaryColor))
        }
    val secondaryColor: OneColor
        get() {
            return if (!config.customTheme) {
                val themeIndex: Int = config.themeIndex
                OneColor(defaultThemes[themeIndex].secondaryColor)
            } else {
                config.secondaryColor
            }
        }
    val darkSecondaryColor: OneColor
        get() {
            return OneColor(darkenColor(secondaryColor))
        }
    val lightSecondaryColor: OneColor
        get() {
            return OneColor(lightenColor(secondaryColor))
        }
    val accentColor: OneColor
        get() {
            return if (!config.useCustomAccentColor) {
                val themeIndex: Int = config.themeIndex
                OneColor(defaultThemes[themeIndex].defaultAccentColor)
            } else {
                config.accentColor
            }
        }
    val darkAccentColor: Color
        get() {
            return darkenColor(accentColor)
        }
    val lightAccentColor: Color
        get() {
            return lightenColor(accentColor)
        }

    private fun darkenColor(color: OneColor): Color {
        return darkenColor(color.toJavaColor())
    }

    private fun darkenColor(color: Color): Color {
        val averageR: Int = (color.red * .761).toInt()
        val averageG: Int = (color.green * .761).toInt()
        val averageB: Int = (color.blue * .761).toInt()
        return Color(averageR, averageG, averageB, color.getTransparency())
    }

    private fun lightenColor(color: OneColor): Color {
        return lightenColor(color.toJavaColor())
    }

    private fun lightenColor(color: Color): Color {
        val averageR: Int = (color.red * .798 + 255 * .202).toInt()
        val averageG: Int = (color.green * .798 + 255 * .202).toInt()
        val averageB: Int = (color.blue * .798 + 255 * .202).toInt()
        return Color(averageR, averageG, averageB, color.getTransparency())
    }

    @Throws(IOException::class)
    fun getBackgroundWithColor(primaryColor: OneColor, secondaryColor: OneColor, accentColor: OneColor): File {
        val primaryColorHex: String = Integer.toHexString(primaryColor.getRGB() and 0xffffff)
        val secondaryColorHex: String = Integer.toHexString(secondaryColor.getRGB() and 0xffffff)
        val accentColorHex: String = Integer.toHexString(accentColor.getRGB() and 0xffffff)
        val folderPath: Path = Paths.get("./config/partly-sane-skies/image_variants/background")
        val fileName = "background-$primaryColorHex-$secondaryColorHex-$accentColorHex.png"
        val filePath: Path = Paths.get("$folderPath/$fileName")

        if (Files.exists(filePath)) {
            return filePath.toFile()
        }

        val backgroundResourceLocation = ResourceLocation("partlysaneskies", "textures/debug_gui_textures/background.png")
        val debugTexture: IResource = minecraft.resourceManager.getResource(backgroundResourceLocation)
        val debugImage: BufferedImage = TextureUtil.readBufferedImage(debugTexture.inputStream)
        folderPath.toFile().mkdirs()
        filePath.toFile().createNewFile()
        replaceColor(debugImage, PRIMARY_DEBUG_COLOR, primaryColor)
        replaceColor(debugImage, SECONDARY_DEBUG_COLOR, secondaryColor)
        replaceColor(debugImage, ACCENT_DEBUG_COLOR, accentColor)
        saveImage(debugImage, filePath)
        return filePath.toFile()
    }

    @Throws(IOException::class)
    fun getButtonWithColor(primaryColor: OneColor, secondaryColor: OneColor, accentColor: OneColor): File {
        val primaryHex: String = Integer.toHexString(primaryColor.getRGB() and 0xffffff)
        val secondaryColorHex: String = Integer.toHexString(secondaryColor.getRGB() and 0xffffff)
        val accentColorHex: String = Integer.toHexString(accentColor.getRGB() and 0xffffff)
        val folderPath: Path = Paths.get("./config/partly-sane-skies/image_variants/button")
        val fileName = "button-$primaryHex-$secondaryColorHex-$accentColorHex.png"
        val filePath: Path = Paths.get("$folderPath/$fileName")

        if (Files.exists(filePath)) {
            return filePath.toFile()
        }
        val buttonResourceLocation: ResourceLocation =
            ResourceLocation("partlysaneskies", "textures/debug_gui_textures/button.png")
        val debugTexture: IResource = minecraft.resourceManager.getResource(buttonResourceLocation)
        val debugImage: BufferedImage = TextureUtil.readBufferedImage(debugTexture.inputStream)
        folderPath.toFile().mkdirs()
        filePath.toFile().createNewFile()
        replaceColor(debugImage, PRIMARY_DEBUG_COLOR, primaryColor)
        replaceColor(debugImage, SECONDARY_DEBUG_COLOR, secondaryColor)
        replaceColor(
            debugImage, ACCENT_DEBUG_COLOR,
            (accentColor)
        )
        saveImage(debugImage, filePath)
        return filePath.toFile()
    }

    @Throws(IOException::class)
    fun getToggleWithColor(primaryColor: OneColor, secondaryColor: OneColor, accentColor: OneColor): File {
        val primaryColorHex: String = Integer.toHexString(primaryColor.getRGB() and 0xffffff)
        val secondaryColorHex: String = Integer.toHexString(secondaryColor.getRGB() and 0xffffff)
        val accentColorHex: String = Integer.toHexString(accentColor.getRGB() and 0xffffff)
        val folderPath: Path = Paths.get("./config/partly-sane-skies/image_variants/toggle")
        val fileName: String = "toggle-$primaryColorHex-$secondaryColorHex-$accentColorHex.png"
        val filePath: Path = Paths.get("$folderPath/$fileName")
        if (Files.exists(filePath)) {
            return filePath.toFile()
        }
        val toggleResourceLocation: ResourceLocation =
            ResourceLocation("partlysaneskies", "textures/debug_gui_textures/toggle.png")
        val debugTexture: IResource = minecraft.resourceManager.getResource(toggleResourceLocation)
        val debugImage: BufferedImage = TextureUtil.readBufferedImage(debugTexture.inputStream)
        folderPath.toFile().mkdirs()
        filePath.toFile().createNewFile()
        replaceColor(debugImage, PRIMARY_DEBUG_COLOR, primaryColor)
        replaceColor(debugImage, SECONDARY_DEBUG_COLOR, secondaryColor)
        replaceColor(
            debugImage, ACCENT_DEBUG_COLOR,
            (accentColor)
        )
        saveImage(debugImage, filePath)
        return filePath.toFile()
    }

    class ButtonData(val image: UIImage, val color: OneColor) {
    }

    class ToggleData(val image: UIImage, selected: Boolean, val color: OneColor) {
        val isSelected: Boolean = selected

    }

    class Theme(val name: String, val primaryColor: Color, val secondaryColor: Color, val defaultAccentColor: Color)
}
